### ISSUE TITLE
SMHI parent class client

### ISSUE THREAD

**Issue by docNord** [Ingenjorsarbete-For-Klimatet/ifk-smhi#29]
## Parent client for the SMHI API. Basic features:

* Join the clients developed together in common framework for getting meterological data
* Get list of available parameters and whether measured or simulated.
* Get closest station a) with specified data b) with any type data 
    * from lat, lon input value
    * from city name input -> openstreetmap ?
    * within radius 
* Get interpolation data

**Comment by mgcth** [comment]
We should probably try to come up with a standard format the user gets data back in from SMHI, maybe a pandas DataFrame?

**Comment by docNord** [comment]
I agree. I think a dataframe is convenient. a datetime-column and a data column, with appropriate column names.

**Comment by docNord** [comment]
We really need to think about how we dynamically change the client object and what to return from functions, I think. When including latitude and longitude searchable functions, if we are to continue with dynamic objects, shouldn't that information be stored in the object in question? And when should the system be reset? 
An alternative would be to use static objects and set up the parameters from them. Also, self.d does not seem like a great name for a variable, should change that...

**Comment by docNord** [comment]
Alright, so I have given this some more thought now. Some comments are also written in [pull request 35](https://github.com/Ingenjorsarbete-For-Klimatet/ifk-smhi/pull/35).

I think that to be able to interpolate data conveniently and also report data missing in a clear manner, the datetime column of the data return variable needs to be promoted to index. Furthermore, I think it would make sense to reindex using the sampling frequency which is readable in the header ("1 gång/dag, kl 6:00" or "1 gång/tim" or similar), which will fill the dataframe with NaN where data is not available. In this way, interpolated data is easily merged into the dataframe at the appropriate position.

I plan to do this in steps: 
1. Promote the datetime to index
2. When interpolation is asked, re-index using the information in the header, available e. g. through regex ´m = re.search('[\d] gång/[^;]*', header);m.group(0)´
3. Find stations within a given distance (using 20km for now) and fill in missing data by merging dataframes.

Once this is done, I will add complementory interpolation steps. A possible choice would be to use data from earlier years, plus of course mesan and strång.

**Comment by mgcth** [comment]
All sounds good to me. Does Strång need interpolation as it's model-based (same with metfcts)?

With regards to 2), are we sure every parameter and station (that returns data) has the sampling frequency in its header? We should either check for that in a bunch of unit tests (that call the API) or handle it through exceptions.

**Comment by docNord** [comment]
I am looking at 2) now. Thus far, it seems they have been good with using the same metadata in all sets, but I am not sure. Probably we should handle it with some sort of exception.

When it comes to Strång and Metfcts, I assume they will not need interpolation - I was thinking of adding that either as the default (if they return the exact values from the stations at all stations, which seems unlikely) or as an option ("use data from smoothing model" or something like that). I was also thinking of a third option, where the station data is used when available and model data is used to cover the gaps.

**Comment by docNord** [comment]
It seems there is not a set way to write measurement frequency - it is always in the same csv cell, but while it is most frequently written as "x gång / y", I have also encountered "en gång per timme" and similar phrasings. I will give this some more thought - perhaps it is easier to parse by reading using csv_read...

**Comment by mgcth** [comment]
> I am looking at 2) now. Thus far, it seems they have been good with using the same metadata in all sets, but I am not sure. Probably we should handle it with some sort of exception.
> 
> When it comes to Strång and Metfcts, I assume they will not need interpolation - I was thinking of adding that either as the default (if they return the exact values from the stations at all stations, which seems unlikely) or as an option ("use data from smoothing model" or something like that). I was also thinking of a third option, where the station data is used when available and model data is used to cover the gaps.

I think metfcts only has data for the past 24 hours so not sure how relevant it is to use that to interpolate historic data.

**Comment by docNord** [comment]
> It seems there is not a set way to write measurement frequency - it is always in the same csv cell, but while it is most frequently written as "x gång / y", I have also encountered "en gång per timme" and similar phrasings. I will give this some more thought - perhaps it is easier to parse by reading using csv_read...

I will use this construct: `data.index.to_series().diff().min()` That will give me the time-delta i need to re_index the dataframe with added NaNs, and it should be fairly general. An additional benefit is that I don't circumvent all the hassle with regular expressions :)

**Comment by docNord** [comment]
> > I am looking at 2) now. Thus far, it seems they have been good with using the same metadata in all sets, but I am not sure. Probably we should handle it with some sort of exception.
> > When it comes to Strång and Metfcts, I assume they will not need interpolation - I was thinking of adding that either as the default (if they return the exact values from the stations at all stations, which seems unlikely) or as an option ("use data from smoothing model" or something like that). I was also thinking of a third option, where the station data is used when available and model data is used to cover the gaps.
> 
> I think metfcts only has data for the past 24 hours so not sure how relevant it is to use that to interpolate historic data.

But metfcts is the forecast, right? We shouldn't use that, no. Let's figure it out when we get there.

**Comment by mgcth** [comment]
> It seems there is not a set way to write measurement frequency - it is always in the same csv cell, but while it is most frequently written as "x gång / y", I have also encountered "en gång per timme" and similar phrasings. I will give this some more thought - perhaps it is easier to parse by reading using csv_read...

Looping over all parameters/stations (is that too comprehensive?) and checking the format? Not a good idea to have such a loop in a test as that would use up all our minutes on GitHub Action (and I don't expect the header to change), but could nonetheless be put inside the test folder.

Could be placed under a notebook folder!

**Comment by mgcth** [comment]
> > > I am looking at 2) now. Thus far, it seems they have been good with using the same metadata in all sets, but I am not sure. Probably we should handle it with some sort of exception.
> > > When it comes to Strång and Metfcts, I assume they will not need interpolation - I was thinking of adding that either as the default (if they return the exact values from the stations at all stations, which seems unlikely) or as an option ("use data from smoothing model" or something like that). I was also thinking of a third option, where the station data is used when available and model data is used to cover the gaps.
> > 
> > 
> > I think metfcts only has data for the past 24 hours so not sure how relevant it is to use that to interpolate historic data.
> 
> But metfcts is the forecast, right? We shouldn't use that, no. Let's figure it out when we get there.

You're correct, its 24-hour forecast!

**Comment by docNord** [comment]
> data.index.to_series().diff().min()

This did not work the way I wanted it to unfortunately, as of course some days the daily measurement will have been performed 7:00 instead of the nominal 6:00, which means the minimum difference will be less than 24h, yielding the re-indexed dataframe useless.

Perhaps I can find a construct using the most frequent time delta somehow.

**Comment by docNord** [comment]
> > data.index.to_series().diff().min()
> 
> This did not work the way I wanted it to unfortunately, as of course some days the daily measurement will have been performed 7:00 instead of the nominal 6:00, which means the minimum difference will be less than 24h, yielding the re-indexed dataframe useless.
> 
> Perhaps I can find a construct using the most frequent time delta somehow.

This works. However, when the first day is measured at 7 instead of at 6, this fails. Also, all days measured at the wrong time of day will fail. I will have to think of something else, unfortunately. I have a construct in mind, but it will probably be messy.

**Comment by docNord** [comment]
Managed to come up with a solution which works, see push above. Somewhat messy, but not as bad as I had feared.

Looking at the example with snow coverage in Kiruna, this is the comparison of interpolated and uninterpolated data:

![snödjup_runt_kiruna_jämförelse](https://user-images.githubusercontent.com/20903303/204136696-0b62ccdd-a1db-4f1f-b16f-a3ae54a1f509.png)

**Comment by docNord** [comment]
With the example of Kiruna, I would have to make the circle considerably larger than 20km in order to fill all the holes, I think (there is a lot of forest once you get outside of the radius covering the mine and the airport, I would assume). I don't know where the radius stops making sense, perhaps that should be left to the user?

**Comment by docNord** [comment]
I get an error on line 17 of test_unit_metobs.py:


```
================================================================= ERRORS ==================================================================
_____________________________________________ ERROR collecting tests/unit/test_unit_metobs.py _____________________________________________
tests\unit\test_unit_metobs.py:17: in <module>
    METOBS_DATA = f.readline().encode("latin1").decode("unicode-escape")
E   UnicodeEncodeError: 'latin-1' codec can't encode character '\u2013' in position 88: ordinal not in range(256)
========================================================= short test summary info =========================================================
ERROR tests/unit/test_unit_metobs.py - UnicodeEncodeError: 'latin-1' codec can't encode character '\u2013' in position 88: ordinal not in...!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

```

**Comment by mgcth** [comment]
Can you try `utf-8` instead of `latin1`?

**Comment by docNord** [comment]
Yes, that works. Thanks!

**Comment by docNord** [comment]
The development included for `MetObs` needs to be extended to the other models as well, notably the "search-by-name-of-place"-functionality. Also, the other interfaces need to be made available from the parent class. @mgcth your thoughts on this?

**Comment by docNord** [comment]
Since `Mesan`, `Strang` and `MetFcts` all accept coordinates as input, all that needs to be included is a preliminary step which gets the latitude and longitude from the city / place to get data from. This is essentially what is implemented already in `find_stations_by_city`:

```python
geolocator = Nominatim(user_agent="ifk-smhi")
loc = geolocator.geocode(city)
lat = loc.latitude
lon = loc.longitude
```

I like the way it is written now, but perhaps I need to kill my darling to be able to generalize this somehow? Thoughts?

**Comment by docNord** [comment]
How about this: Include `Mesan`, `Strang` and `MetFcts` functions of the type `get_point` and `get_multipoint`, but include variable call structures such that it could either be called using `lat`, `lon` or with a string `city`?

Like:
```python
data = get_metfcts_point('Gothenburg')
```

Could be a useful addition, couldn't it?

**Comment by docNord** [comment]
We have a new issue for the continuation work - most of the issue as initially formulated here has been solved. I am closing now.
